File: wrappers/dart/lib/src/orm/filter.dart
/// Typed filter expressions for the query builder.
///
/// A [Col] is a typed reference to a column; its comparison methods build a
/// [Filter]. Filters compose with `&` (AND) and `|` (OR) and lower to a
/// parameterised SGQL `WHERE` fragment — values become `$1`, `$2`, … bindings,
/// never interpolated, so quoting and types are always correct.
library;

/// Accumulates positional parameters while a filter renders to SGQL.
class SqlContext {
  final List<Object?> params = [];

  /// Register [value] and return its placeholder (`$1`, `$2`, …).
  String placeholder(Object? value) {
    params.add(value);
    return '\$${params.length}';
  }
}

/// A boolean predicate over a row. Compose with `&` / `|`.
abstract class Filter {
  const Filter();

  /// Render to an SGQL boolean expression, binding literals via [ctx].
  String render(SqlContext ctx);

  Filter operator &(Filter other) => AndFilter(this, other);
  Filter operator |(Filter other) => OrFilter(this, other);
}

/// A typed column reference. Comparison methods are type-checked: `price.lt('x')`
/// on a `Col<int>` is a compile error.
class Col<T> {
  final String name;
  const Col(this.name);

  Filter eq(T value) => CompareFilter(name, '=', value);
  Filter neq(T value) => CompareFilter(name, '!=', value);
  Filter lt(T value) => CompareFilter(name, '<', value);
  Filter lte(T value) => CompareFilter(name, '<=', value);
  Filter gt(T value) => CompareFilter(name, '>', value);
  Filter gte(T value) => CompareFilter(name, '>=', value);

  /// `name BETWEEN lo AND hi` (inclusive).
  Filter between(T lo, T hi) => BetweenFilter(name, lo, hi);
}

/// `name OP $n`.
class CompareFilter extends Filter {
  final String column;
  final String op;
  final Object? value;
  const CompareFilter(this.column, this.op, this.value);

  @override
  String render(SqlContext ctx) => '$column $op ${ctx.placeholder(value)}';
}

/// `name BETWEEN $lo AND $hi`.
class BetweenFilter extends Filter {
  final String column;
  final Object? lo;
  final Object? hi;
  const BetweenFilter(this.column, this.lo, this.hi);

  @override
  String render(SqlContext ctx) =>
      '$column BETWEEN ${ctx.placeholder(lo)} AND ${ctx.placeholder(hi)}';
}

/// `(a AND b)`.
class AndFilter extends Filter {
  final Filter a;
  final Filter b;
  const AndFilter(this.a, this.b);

  @override
  String render(SqlContext ctx) => '(${a.render(ctx)} AND ${b.render(ctx)})';
}

/// `(a OR b)`.
class OrFilter extends Filter {
  final Filter a;
  final Filter b;
  const OrFilter(this.a, this.b);

  @override
  String render(SqlContext ctx) => '(${a.render(ctx)} OR ${b.render(ctx)})';
}

File: wrappers/kotlin/orm/src/main/kotlin/life/sekejap/Filter.kt
package life.sekejap

/** Accumulates positional parameters while a filter renders to SGQL. Params are
 *  serialized with a hand-rolled JSON writer (no kotlinx allocation per call) —
 *  this is the write-path hot loop for update/delete. */
class SqlContext {
    private val params = mutableListOf<Any?>()

    fun placeholder(value: Any?): String {
        params.add(value)
        return "\$${params.size}"
    }

    fun paramsJson(): String = buildString {
        append('[')
        params.forEachIndexed { i, v ->
            if (i > 0) append(',')
            appendJson(v)
        }
        append(']')
    }

    private fun StringBuilder.appendJson(v: Any?) {
        when (v) {
            null -> append("null")
            is Boolean, is Int, is Long, is Double, is Float -> append(v.toString())
            is String -> { append('"'); appendEscaped(v); append('"') }
            else -> { append('"'); appendEscaped(v.toString()); append('"') }
        }
    }

    private fun StringBuilder.appendEscaped(s: String) {
        for (c in s) when (c) {
            '"' -> append("\\\"")
            '\\' -> append("\\\\")
            '\n' -> append("\\n")
            '\r' -> append("\\r")
            '\t' -> append("\\t")
            else -> append(c)
        }
    }
}

/** A boolean predicate over a row. Compose with [and] / [or]. */
sealed class Filter {
    abstract fun render(ctx: SqlContext): String
    infix fun and(other: Filter): Filter = AndFilter(this, other)
    infix fun or(other: Filter): Filter = OrFilter(this, other)
}

/** A typed column reference — the receiver in `where { it.field eq x }`. */
class Col<T>(val name: String) {
    infix fun eq(value: T): Filter = Compare(name, "=", value)
    infix fun neq(value: T): Filter = Compare(name, "!=", value)
    infix fun lt(value: T): Filter = Compare(name, "<", value)
    infix fun lte(value: T): Filter = Compare(name, "<=", value)
    infix fun gt(value: T): Filter = Compare(name, ">", value)
    infix fun gte(value: T): Filter = Compare(name, ">=", value)
    fun between(lo: T, hi: T): Filter = Between(name, lo, hi)
}

class Compare(val column: String, val op: String, val value: Any?) : Filter() {
    override fun render(ctx: SqlContext) = "$column $op ${ctx.placeholder(value)}"
}

class Between(val column: String, val lo: Any?, val hi: Any?) : Filter() {
    override fun render(ctx: SqlContext) =
        "$column BETWEEN ${ctx.placeholder(lo)} AND ${ctx.placeholder(hi)}"
}

class AndFilter(val a: Filter, val b: Filter) : Filter() {
    override fun render(ctx: SqlContext) = "(${a.render(ctx)} AND ${b.render(ctx)})"
}

class OrFilter(val a: Filter, val b: Filter) : Filter() {
    override fun render(ctx: SqlContext) = "(${a.render(ctx)} OR ${b.render(ctx)})"
}

/** A WGS84 longitude/latitude point for spatial predicates. */
data class GeoPoint(val lon: Double, val lat: Double) {
    fun toSql(): String = "POINT($lon $lat)"
}

/** Vector similarity used for ranking (higher is nearer, ranked DESC). */
enum class VectorMetric(val fn: String) { COSINE("VECTOR_COSINE"), DOT("VECTOR_DOT") }
